File: r/bioviz/tests/testthat/test-treemap.R
test_that("treemap() builds an htmlwidget with the expected payload", {
  df <- data.frame(
    id = c("root", "P1", "P2", "g1", "g2", "g3"),
    parent = c(NA, "root", "root", "P1", "P1", "P2"),
    value = c(0, 0, 0, 3, 5, 2),
    label = c("All", "Pathway 1", "Pathway 2", "Gene 1", "Gene 2", "Gene 3")
  )
  w <- treemap(df,
    tile = "binary", padding_inner = 2,
    color_by = "value", colormap = "rdbu", label_min_size = 40
  )

  expect_s3_class(w, "htmlwidget")
  expect_equal(w$x$data$columns$id, df$id)
  # NA parent (root) is encoded as an empty string.
  expect_equal(w$x$data$columns$parent[1], "")
  expect_equal(w$x$data$columns$parent[4], "P1")
  expect_equal(w$x$data$columns$value, df$value)
  expect_equal(w$x$data$meta$labels, df$label)
  expect_equal(w$x$options$tile, "binary")
  expect_equal(w$x$options$paddingInner, 2)
  expect_equal(w$x$options$colorBy, "value")
  expect_equal(w$x$options$colormap, "rdbu")
  expect_equal(w$x$options$labelMinSize, 40)
})

test_that("treemap() defaults value/label sensibly when absent", {
  w <- treemap(data.frame(id = c("a", "b"), parent = c(NA, "a")))
  expect_equal(w$x$data$columns$value, c(0, 0))
  expect_null(w$x$data$meta$labels)
  expect_equal(w$x$options$tile, "squarify")
  expect_equal(w$x$options$colorBy, "parent")
})

test_that("treemap() validates its input", {
  expect_error(treemap(list(id = 1, parent = 2)), "must be a data frame")
  expect_error(treemap(data.frame(a = 1, b = 2)), "must contain columns")
})

test_that("treemap() rejects unknown option values", {
  df <- data.frame(id = "a", parent = NA)
  expect_error(treemap(df, tile = "spiral"))
  expect_error(treemap(df, colormap = "jet"))
})

test_that("treemap() rejects duplicate node ids", {
  df <- data.frame(id = c("a", "a"), parent = c(NA, "a"))
  expect_error(treemap(df), "duplicate node id\\(s\\): a")
})

test_that("treemap() requires exactly one root", {
  two_roots <- data.frame(id = c("a", "b"), parent = c(NA, NA))
  expect_error(treemap(two_roots), "treemap requires exactly one root")
  no_root <- data.frame(id = c("a", "b"), parent = c("b", "a"))
  expect_error(treemap(no_root), "treemap requires exactly one root")
})

test_that("treemap() rejects a parent not found among ids", {
  df <- data.frame(id = c("root", "g1"), parent = c(NA, "nope"))
  expect_error(treemap(df), "parent\\(s\\) not found among ids: nope")
})

test_that("treemap() rejects a factor value column", {
  df <- data.frame(
    id = c("root", "g1"), parent = c(NA, "root"),
    value = factor(c("0", "3"))
  )
  expect_error(treemap(df), "`value` must be numeric")
})

test_that("treemap() rejects empty input", {
  df <- data.frame(id = character(0), parent = character(0))
  expect_error(treemap(df), "`data` has no rows/cells")
})

test_that("treemap() forwards a theme override", {
  df <- data.frame(id = c("a", "b"), parent = c(NA, "a"))
  w <- treemap(df, theme = list(background = "#222"))
  expect_equal(w$x$options$theme$background, "#222")
  expect_null(treemap(df)$x$options$theme)
})

test_that("treemap() rejects a cycle disjoint from the root", {
  # One valid root, but a and b point at each other: passes the one-root and
  # orphan checks yet crashes d3.stratify in the browser.
  df <- data.frame(id = c("root", "a", "b"), parent = c(NA, "b", "a"))
  expect_error(treemap(df), "cycle")
})

test_that("treemap() treats a literal \"NA\" parent as the root", {
  # Matches the JS/Python contract: both "" and "NA" mark the root.
  df <- data.frame(
    id = c("root", "g1"), parent = c("NA", "root"),
    stringsAsFactors = FALSE
  )
  w <- treemap(df)
  expect_equal(w$x$data$columns$parent[1], "")
})

File: python/tests/test_treemap.py
import numpy as np
import pytest

from plotomics import Treemap


def _sample():
    return {
        "id": ["root", "P1", "P2", "g1", "g2", "g3"],
        "parent": [None, "root", "root", "P1", "P1", "P2"],
        "value": [0, 0, 0, 3, 5, 2],
        "label": ["All", "Pathway 1", "Pathway 2", "Gene 1", "Gene 2", "Gene 3"],
    }


def test_treemap_builds_traits_from_dict():
    w = Treemap(
        _sample(),
        tile="binary",
        padding_inner=2,
        color_by="value",
        colormap="rdbu",
        label_min_size=40,
    )

    assert w.data["columns"]["id"] == _sample()["id"]
    # None parent (root) becomes an empty string.
    assert w.data["columns"]["parent"][0] == ""
    assert w.data["columns"]["parent"][3] == "P1"
    assert w.data["meta"]["labels"][1] == "Pathway 1"

    assert w.options["tile"] == "binary"
    assert w.options["paddingInner"] == 2
    assert w.options["colorBy"] == "value"
    assert w.options["colormap"] == "rdbu"
    assert w.options["labelMinSize"] == 40

    # value packed as one float32 column of length 6 (consistent with all
    # other components' packed columns).
    names = {c["name"] for c in w.schema["columns"]}
    assert names == {"value"}
    assert w.schema["columns"][0]["dtype"] == "float32"
    assert len(w.buffer) == 6 * 4
    assert np.frombuffer(w.buffer, dtype=np.float32).tolist() == [0, 0, 0, 3, 5, 2]


def test_treemap_defaults():
    w = Treemap({"id": ["a", "b"], "parent": [None, "a"]})
    assert w.options["tile"] == "squarify"
    assert w.options["colorBy"] == "parent"
    assert w.options["colormap"] == "viridis"
    # value defaults to zeros when absent (float32).
    assert np.frombuffer(w.buffer, dtype=np.float32).tolist() == [0.0, 0.0]
    assert "labels" not in w.data["meta"]


def test_treemap_theme_in_options():
    w = Treemap({"id": ["a", "b"], "parent": [None, "a"]}, theme={"background": "#000"})
    assert w.options["theme"] == {"background": "#000"}


def test_treemap_rejects_duplicate_ids():
    with pytest.raises(ValueError, match="duplicate node id"):
        Treemap({"id": ["a", "a"], "parent": [None, "a"]})


def test_treemap_requires_exactly_one_root():
    # Two roots (two empty parents).
    with pytest.raises(ValueError, match="exactly one root"):
        Treemap({"id": ["a", "b"], "parent": [None, None]})
    # Zero roots (every node has a parent).
    with pytest.raises(ValueError, match="exactly one root"):
        Treemap({"id": ["a", "b"], "parent": ["b", "a"]})


def test_treemap_rejects_orphan_parent():
    with pytest.raises(ValueError, match="not found among ids"):
        Treemap({"id": ["root", "g1"], "parent": [None, "missing"]})


def test_treemap_rejects_empty_data():
    with pytest.raises(ValueError, match="at least one row"):
        Treemap({"id": [], "parent": []})


def test_treemap_rejects_cycle():
    # One valid root, but a and b point at each other: a cycle disjoint from the
    # root that passes the one-root / orphan checks yet crashes d3.stratify.
    with pytest.raises(ValueError, match="cycle"):
        Treemap({"id": ["root", "a", "b"], "parent": [None, "b", "a"]})


def test_treemap_rejects_self_parent():
    with pytest.raises(ValueError, match="cycle"):
        Treemap({"id": ["root", "a"], "parent": [None, "a"]})


def test_treemap_rejects_non_numeric_value():
    with pytest.raises(ValueError, match="must be numeric"):
        Treemap(
            {"id": ["root", "a"], "parent": [None, "root"], "value": ["x", "y"]}
        )


def test_treemap_requires_id_and_parent():
    with pytest.raises(ValueError, match="id` and `parent`"):
        Treemap({"id": ["a", "b"]})


def test_treemap_accepts_pandas_dataframe():
    pd = pytest.importorskip("pandas")
    df = pd.DataFrame(_sample())
    w = Treemap(df)
    assert w.data["columns"]["id"] == _sample()["id"]
    # pandas turns None into NaN; both normalize to the empty-string root marker.
    assert w.data["columns"]["parent"][0] == ""
    assert w.data["meta"]["labels"] == _sample()["label"]
